### PR TITLE
Updat ZoneTransaction to support operations when firewalld is offline

### LINKED ISSUE #398: ZoneTransactions fail with 'FirewallConfig' object is not callable` when firewalld is offline
##### SUMMARY
When attempting to create a new zone and firewalld is currently stopped, the task fails with the message:
`ERROR: Exception caught: 'FirewallConfig' object is not callable`

On review, it appears that the python APIs for online and offline changes are entirely different and all methods in ZoneTransaction naively use the online API only. Reviewing history, the implementation appears unchanged at least as far back as March 2020 when the module was in community.general.

The required changes appear fairly straightforward and I'm open to raising a pull request; however, I would appreciate some guidance on the scope of testing.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible.posix.firewalld

##### ANSIBLE VERSION

```paste below
2.12.7 - 2.13.6
```

##### COLLECTION VERSION

```paste below
1.4.0
```

##### CONFIGURATION

```paste below
N/A
```

##### OS / ENVIRONMENT

ArchLinux (latest)
Firewalld 1.2.1-1

##### STEPS TO REPRODUCE

```yaml
- name: 'Zone example - Create new zone'
  ansible.posix.firewalld:
    zone: "example"
    state: "present"
    permanent: Yes
```

##### EXPECTED RESULTS

A new zone named example is created

##### ACTUAL RESULTS

```paste below
The full traceback is:
  File "/tmp/ansible_ansible.posix.firewalld_payload_04lptorx/ansible_ansible.posix.firewalld_payload.zip/ansible_collections/ansible/posix/plugins/module_utils/firewalld.py", line 111, in action_handler
    return action_func(*action_func_args)
  File "/tmp/ansible_ansible.posix.firewalld_payload_04lptorx/ansible_ansible.posix.firewalld_payload.zip/ansible_collections/ansible/posix/plugins/modules/firewalld.py", line 678, in get_enabled_permanent
fatal: [boot]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "icmp_block": null,
            "icmp_block_inversion": null,
            "immediate": false,
            "interface": null,
            "masquerade": null,
            "offline": null,
            "permanent": true,
            "port": null,
            "port_forward": null,
            "rich_rule": null,
            "service": null,
            "source": null,
            "state": "present",
            "target": null,
            "timeout": 0,
            "zone": "example"
        }
    },
    "msg": "ERROR: Exception caught: 'FirewallConfig' object is not callable"
}
```

### PR DESCRIPTION
**Pull request by gnfzdz**
##### SUMMARY

Fixes #398 by checking the flag self.fw_offline and calling the offline specific APIs when the flag is true.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible.posix.firewalld

##### ADDITIONAL INFORMATION
When attempting to add or remove zones, the ansible.posix.firewalld module would always call APIs as if it was online.

Specifically, the ZoneTransaction class assumed that self.fw  was a [FirewallClient](https://github.com/firewalld/firewalld/blob/e1407b44abc1618165accca2744226e3f0f74627/src/firewall/client.py#L2835), but when the daemon is offline, it is instead either a [Firewall](https://github.com/firewalld/firewalld/blob/e1407b44abc1618165accca2744226e3f0f74627/src/firewall/core/fw.py#L70) or [Firewall_test](https://github.com/firewalld/firewalld/blob/44200d0f508a990c5dfff9f480a6206ec507e229/src/firewall/core/fw_test.py#L55) based on the version of firewalld installed.

See #398  for additional background.

###### Sample task
```yaml
- name: 'Zone example - Create new zone'
  ansible.posix.firewalld:
    zone: "example"
    state: "present"
    permanent: Yes
```

###### Before
```
The full traceback is:
  File "/tmp/ansible_ansible.posix.firewalld_payload_04lptorx/ansible_ansible.posix.firewalld_payload.zip/ansible_collections/ansible/posix/plugins/module_utils/firewalld.py", line 111, in action_handler
    return action_func(*action_func_args)
  File "/tmp/ansible_ansible.posix.firewalld_payload_04lptorx/ansible_ansible.posix.firewalld_payload.zip/ansible_collections/ansible/posix/plugins/modules/firewalld.py", line 678, in get_enabled_permanent
fatal: [boot]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "icmp_block": null,
            "icmp_block_inversion": null,
            "immediate": false,
            "interface": null,
            "masquerade": null,
            "offline": null,
            "permanent": true,
            "port": null,
            "port_forward": null,
            "rich_rule": null,
            "service": null,
            "source": null,
            "state": "present",
            "target": null,
            "timeout": 0,
            "zone": "example"
        }
    },
    "msg": "ERROR: Exception caught: 'FirewallConfig' object is not callable"
}
```

###### After
```paste below
changed: [boot] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "icmp_block": null,
            "icmp_block_inversion": null,
            "immediate": false,
            "interface": null,
            "masquerade": null,
            "offline": null,
            "permanent": true,
            "port": null,
            "port_forward": null,
            "rich_rule": null,
            "service": null,
            "source": null,
            "state": "present",
            "target": null,
            "timeout": 0,
            "zone": "example"
        }
    },
    "msg": "Permanent operation, Added zone example, Changed zone example to present, (offline operation: only on-disk configs were altered)"
```

### CODE REVIEW

**Comment by Akasurde** [comment]
```
2022-11-30 23:17:05.579138 | controller | ceeff5403d9e4c18d53dc913655ca0d5e5b54f2e78c868b18d901909000fd325
2022-11-30 23:17:05.595494 | controller | Run command: podman image inspect quay.io/ansible/default-test-container:7.4.0
2022-11-30 23:17:05.689826 | controller | Run command: podman network inspect podman
2022-11-30 23:17:05.733372 | controller | Run command: podman run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --systemd always --cgroupns private -dt --ulimit nofile=10240 --name ansible-test-controller-mSpD8lZM --network podman quay.io/ansible/default-test-container:7.4.0
2022-11-30 23:17:06.402912 | controller | ERROR: Command "podman run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --systemd always --cgroupns private -dt --ulimit nofile=10240 --name ansible-test-controller-mSpD8lZM --network podman quay.io/ansible/default-test-container:7.4.0" returned exit status 126.
2022-11-30 23:17:06.402945 | controller | >>> Standard Error
2022-11-30 23:17:06.402954 | controller | Error: netavark: code: 3, msg: modprobe: ERROR: could not insert 'ip_tables': Operation not permitted
2022-11-30 23:17:06.402961 | controller | iptables v1.8.7 (legacy): can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
2022-11-30 23:17:06.402967 | controller | Perhaps iptables or your kernel needs to be upgraded.[0m
2022-11-30 23:17:06.402973 | controller | WARNING: Failed to run docker image "{image}". Waiting a few seconds before trying again.
2022-11-30 23:17:06.402979 | controller | Run command: podman stop --time 0 ansible-test-controller-mSpD8lZM
2022-11-30 23:17:06.486167 | controller | Run command: podman rm ansible-test-controller-mSpD8lZM
2022-11-30 23:17:09.585131 | controller | Run command: podman run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --systemd always --cgroupns private -dt --ulimit nofile=10240 --name ansible-test-controller-mSpD8lZM --network podman quay.io/ansible/default-test-container:7.4.0
2022-11-30 23:17:09.816186 | controller | ERROR: Command "podman run --tmpfs /tmp:exec --tmpfs /run:exec --tmpfs /run/lock --systemd always --cgroupns private -dt --ulimit nofile=10240 --name ansible-test-controller-mSpD8lZM --network podman quay.io/ansible/default-test-container:7.4.0" returned exit status 126.
2022-11-30 23:17:09.816221 | controller | >>> Standard Error
2022-11-30 23:17:09.816233 | controller | Error: netavark: code: 3, msg: modprobe: ERROR: could not insert 'ip_tables': Operation not permitted
2022-11-30 23:17:09.816243 | controller | iptables v1.8.7 (legacy): can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
2022-11-30 23:17:09.816251 | controller | Perhaps iptables or your kernel needs to be upgraded.[0m
2022-11-30 23:17:09.816259 | controller | WARNING: Failed to run docker image "{image}". Waiting a few seconds before trying again.
2022-11-30 23:17:09.816266 | controller | Run command: podman stop --time 0 ansible-test-controller-mSpD8lZM
2022-11-30 23:17:09.871083 | controller | Run command: podman rm ansible-test-controller-mSpD8lZM
2022-11-30 23:17:13.864191 | controller | ERROR: Host DockerConfig(python=NativePythonConfig(version='3.11', path='/usr/bin/python3.11'), name='default', image='quay.io/ansible/default-test-container:7.4.0', memory=None, privileged=False, seccomp='default', cgroup=CGroupVersion.V1_V2, audit=AuditMode.REQUIRED) job failed:
2022-11-30 23:17:13.864239 | controller | Failed to run docker image "quay.io/ansible/default-test-container:7.4.0".
2022-11-30 23:17:13.864249 | controller | FATAL: Host job(s) failed. See previous error(s) for details.
2022-11-30 23:17:14.132006 | controller | ERROR
2022-11-30 23:17:14.132204 | controller | {
2022-11-30 23:17:14.132252 | controller |   "delta": "0:00:58.494196",
2022-11-30 23:17:14.132287 | controller |   "end": "2022-11-30 23:17:13.903209",
2022-11-30 23:17:14.132318 | controller |   "msg": "non-zero return code",
2022-11-30 23:17:14.132349 | controller |   "rc": 1,
2022-11-30 23:17:14.132406 | controller |   "start": "2022-11-30 23:16:15.409013"
2022-11-30 23:17:14.132453 | controller | }
```

Failures are related to Container Image that is being used.

**Comment by gnfzdz** [comment]
I reviewed the build failures on RHEL 9 (devel & 2.14). This failure is also present in the scheduled builds on the main branch, so appears unrelated to this pull request.

The failure seems to occur when the daemon is offline and multiple overlapping ports have already been defined within the zone. It looks like [this commit to firewalld](https://github.com/firewalld/firewalld/commit/9592097b94622adbc57246a9af9e9623b1944532?diff=split) made a change so that when loading the config file from disk, ports/port ranges that are entirely inclusive to an already loaded range are ignored (with a warning). When checking if the port is already enabled, the Ansible Firewalld module is currently only checking for an exact match.

I'm open to contributing a solution for this issue as well, but I think it should be a separate PR. Let me know how best to proceed.

**Comment by Akasurde** [comment]
> I'm open to contributing a solution for this issue as well, but I think it should be a separate PR. Let me know how best to proceed.

👍🏼 I will check if someone else encountered these failures in CI. cc @goneri Have you seen these errors before?

**Comment by goneri** [comment]
recheck

**Comment by goneri** [comment]
> > I'm open to contributing a solution for this issue as well, but I think it should be a separate PR. Let me know how best to proceed.
> 
> 👍🏼 I will check if someone else encountered these failures in CI. cc @goneri Have you seen these CI failures before?

We had a similar problem recently with Matt Clay. We run the tests in a non-root podman that cannot load kernel module. We need to manually load nft/iptable modules before.

**Comment by goneri** [comment]
recheck

**Comment by Akasurde** [comment]
@goneri Thanks a lot for taking a look into this.

**Comment by gnfzdz** [comment]
@Akasurde @goneri The build still appears to be failing both here and in the main branch due to the issue I mentioned with RHEL 9 and the upstream changes to firewalld. It looks like you were working through some other failures after my initial comment and I just want to confirm that this second issue wasn't lost in the confusion (especially as my PR is also making changes to the firewalld module).

@mattclay we briefly discussed this in Matrix last week as well.

To rephrase my last post, firewalld made a change that if a port range is already allowed, then an attempt to add another port within that range will be ignored. The ansible.posix.firewalld integration test is specifically testing this scenario: 5500-6950 are already allowed and attempts to also allow 6900.

Assuming, this change in behavior was intentional then the Ansible module's integration test is no longer valid. The ansible.posix.firewalld module would still require a change when testing if a port/port range is already open as it's currently just checking for an exact match.

Hopefully this doesn't come across as pestering. I'm not actually in a rush to get this merged, I just want to make sure the issue wasn't lost in the confusion. Let me know if there's anything I can do to help get it merged. Thanks!

**Comment by goneri** [comment]
I only see the following error in the `ansible-test-sanity-docker-stable-*` jobs:

```
ERROR: tests/integration/targets/firewalld/tasks/zone_test_cases.yml:1:1: empty-lines: too many blank lines (1 > 0)
```

**Comment by gnfzdz** [comment]
@goneri  Thanks, I actually missed that failure as I was originally looking at the Azure pipeline failures linked at the bottom of the PR, not the Zuul failures linked a bit further up (which I'd interpreted as old). I just pushed a fix to remove the empty line mentioned in your comment.

These are the failures that I was addressing in my above comments. Are these non-voting as well? [my pr](https://dev.azure.com/ansible/ansible.posix/_build/results?buildId=61842&view=logs&j=d16e35df-7435-5602-c998-91f475d35ec9&t=6c6f8326-7cd0-58e5-387d-db4a70ae8e77&l=1813 ) / [main](https://dev.azure.com/ansible/ansible.posix/_build/results?buildId=62224&view=logs&j=d16e35df-7435-5602-c998-91f475d35ec9&t=6c6f8326-7cd0-58e5-387d-db4a70ae8e77&l=2213)
```
02:03 TASK [firewalld : assert firewalld port range test permanent enabled worked] ***
02:03 task path: /root/ansible_collections/ansible/posix/tests/output/.tmp/integration/firewalld-pojiwtca-ÅÑŚÌβŁÈ/tests/integration/targets/firewalld/tasks/port_test_cases.yml:12
02:03 fatal: [testhost]: FAILED! => {
02:03     "assertion": "result is changed",
02:03     "changed": false,
02:03     "evaluated_to": false,
02:03     "msg": "Assertion failed"
02:03 }
```

**Comment by gnfzdz** [comment]
@Akasurde @goneri Is there anything else I need to do before this PR can be merged? I have a few open PRs related to the Firewalld module with ansible/gate waiting for status to be reported.

**Review by goneri — APPROVED** [review]

**Comment by goneri** [comment]
I'm not the maintainer of the collection but the PR looks good to me.

**Comment by gnfzdz** [comment]
@goneri Thanks for the approval! It doesn't look like there's anything configured for this module in [BOTMETA](https://github.com/ansible-collections/ansible.posix/blob/main/.github/BOTMETA.yml). @maxamillion as the original author, are you still also the maintainer for this module?

**Pull request merged**
